### ISSUE TITLE
[HOLD for payment 2024-08-19] [$250] Concierge - Error page when new account navigate to Concierge via Link

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#44235]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.1-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4661673&group_by=cases:section_id&group_order=asc&group_id=229066
**Email or phone of affected tester (no customers):** gocemate+a388@gmail.com
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
1. Log out of NewDot
2. Navigate to https://staging.new.expensify.com/concierge
3. Sign-up with a new account
4. Finish the onboarding flow

## Expected Result:
User should navigate to the concierge chat

## Actual Result:
Error page appear when new user navigates to concierge link after finishing the onboarding flow

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/01b8e0ac-9629-42c5-ab76-1f1c48c79f34

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017c459a6423831994</li>
        <li>Upwork Job ID: 1805205450283630802</li>
        <li>Last Price Increase: 2024-07-22</li>
<li>Automatic offers: </li>
<ul>
<li>ahmedGaber93 | Reviewer | 103318767</li>
<li>NJ-2020 | Contributor | 103318769</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @isabelastisser</details>

**Comment by izarutskaya** [comment]
@isabelastisser I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors.

**Comment by izarutskaya** [comment]
Production

https://github.com/Expensify/App/assets/115492554/f9010b61-e9c3-473d-a686-afa56a55f4e5

**Comment by mountiny** [comment]
demoting since this is an edge case and the user is not blocked from using the app, looking for proposals to fix this though

**Comment by jnowakow** [comment]
Hello, I'm from SoftwareMansion, an expert agency, and I can work on this. 
It seems that the problem is cause by `Member.clearWorkspaceOwnerChangeFlow(policyID);` being called right before opening modal. Probably this causes page to re-render

**Comment by jnowakow** [comment]
I commented it out and it help on one workspace but when I navigate to other report I still can observe the problem. I'm looking for other cause

**Comment by pankajsoftyoi** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
 - The user should stay in concierge chat after finishing the onboarding steps.
### What is the root cause of that problem?
 - Redirection issue is due to a check on the account ID. A/B testing code was used in the condition to redirect to the error page. 
### What changes do you think we should make in order to solve the problem?


https://github.com/Expensify/App/blob/fd4987fca779837cb0823091d547625c510b03d8/src/pages/OnboardingPersonalDetails/BaseOnboardingPersonalDetails.tsx#L74-L78

We need to remove A/B testing code and redirect it to the concierge chat always.
if we remove above condition and just replace it with `Report.navigateToConciergeChat();` then it will resolve redirection issue.

### What alternative solutions did you explore? (Optional)

N/A

**Comment by pankajsoftyoi** [comment]
Contributor details
Your Expensify account email: pankaj.bhalala@softyoi.com
Upwork Profile Link: https://www.upwork.com/freelancers/~012986a83dc460e833

**Comment by ahmedGaber93** [comment]
@pankajsoftyoi Thanks for the proposal.

I don't think your root cause is correct, this code shouldn't display error page

https://github.com/Expensify/App/blob/fd4987fca779837cb0823091d547625c510b03d8/src/pages/OnboardingPersonalDetails/BaseOnboardingPersonalDetails.tsx#L74-L78

also the Error page displayed before execute the above code (before click "continue" button)

![Screenshot 2024-06-25 at 10 24 26 PM](https://github.com/Expensify/App/assets/41129870/305b167f-3120-4edb-9d9c-deab1ac09151)

**Comment by NJ-2020** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Error page when navigating to Concierge via a link with a new account

### What is the root cause of that problem?
When we navigate to Concierge via a link, this function is called
https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/pages/ConciergePage.tsx#L43
But this function run before OpenApp completed. So when we check if we do not have a Concierge report id, it will return true, and it will run the navigateToAndOpenReport function
https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/libs/actions/Report.ts#L2014-L2028 
And when we find existing Concierge report chat, the reports value of getChatByParticipants function return undefined, because we run before OpenApp competed, so when we check existing Concierge report chat, it will return undefined
https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/libs/actions/Report.ts#L973-L975
https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/libs/ReportUtils.ts#L5526
So when we check if there is no existing Concierge report chat, it will create a report chat for Concierge but only in the optimistic data and do not push to the backend, and when we try to open the report with the report id from the optimistic data, the backend couldn't find that report with the optimistic report id, so it will return error
https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/libs/actions/Report.ts#L999
### What changes do you think we should make in order to solve the problem?
We can run the `navigateToConciergeChat` function when isLoading is false, meaning OpenApp is completed, or any other way that check the OpenApp is completed
### What alternative solutions did you explore? (Optional)

**Comment by NJ-2020** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/44235#issuecomment-2194648301)

**Comment by isabelastisser** [comment]
@ahmedGaber93, please review the updated proposals above. thanks!

**Comment by ahmedGaber93** [comment]
Reviewing today

**Comment by ahmedGaber93** [comment]
@NJ-2020 Thanks for the proposal.

Your solution not work with me, could you please mention the exact changes you want to do?
Also, I am not sure about the root cause yet, `openApp` API that back with the backend Concierge report is completely finished before calling `openReport` see second 26 in the video


https://github.com/Expensify/App/assets/41129870/1448141c-570f-4723-8de2-981d6912423d

**Comment by NJ-2020** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/44235#issuecomment-2194648301)

**Comment by NJ-2020** [comment]
@ahmedGaber93
Kindly, please took my updated proposal

I've tried to solve it in the API, waiting for OnyxData completely applied to Onyx then we can apply the finallyData but still it doesn't work

Many Thanks

**Comment by madmax330** [comment]
Still waiting on a good proposal, @ahmedGaber93 can you check the above proposal?

**Comment by ahmedGaber93** [comment]
I am checking, and I will put feedback shortly.

**Comment by ahmedGaber93** [comment]
After retest the latest main, I found [Report.navigateToConciergeChat](https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/pages/ConciergePage.tsx#L43) is not called, this because the function return early by false. It shouldn't return early because `isUnmounted.current` should be false while the `ConciergePage` is not unmounted.

https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/pages/ConciergePage.tsx#L40-L42

To still able to reproduce the issue, you can comment the line below

https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/pages/ConciergePage.tsx#L50-L55

I think we need to fix it before move forward here.

**Comment by ahmedGaber93** [comment]
@NJ-2020 Thanks for the update.

I see your solution work good in your video, but I think your root cause is not complete.

https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/libs/actions/Report.ts#L2022

When we navigate to concierge and `conciergeChatReportID` not found [here](https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/libs/actions/Report.ts#L2014), we pass `participantAccountIDs` to backend and backend return with `preexistingReportID` (conciergeChatReportID) and the app should do that
1. remove the optimistic report [here](https://github.com/Expensify/App/blob/81abfeca6697ee1f2cda38f960a0d913ee62856b/src/libs/actions/Report.ts#L1297)
2. redirect to pre-exist chat [here](https://github.com/Expensify/App/blob/81abfeca6697ee1f2cda38f960a0d913ee62856b/src/libs/actions/Report.ts#L1304)

In this issue case, The optimistic report already removed, but the app didn't redirect to pre-exist chat because this [condition](https://github.com/Expensify/App/blob/81abfeca6697ee1f2cda38f960a0d913ee62856b/src/libs/actions/Report.ts#L1300) will return false as the current route now is `onboarding/purpose`, and the removed report will cause the error page.


___

about the @NJ-2020 solution that will display loading report skeleton and wait app to complete loading and complete `onboarding/purpose`and `ConciergePage` come back to focus then navigate to concierge (screen is focused and app is loaded), 

the solution should work fine because we navigate to `onboarding/purpose` after app complete loading [here](https://github.com/Expensify/App/blob/11982271301889acf5116d06722d72931c941d3f/src/libs/Navigation/AppNavigator/createCustomBottomTabNavigator/BottomTabBar/index.tsx#L60-L62) so condition (ConciergePage screen is focused and app is loaded) will be false util `onboarding/purpose` close and `ConciergePage` come back to focus.

By I think `waitForAppLoaded` is used only on E2E tests, so we can replace it with this code [here](https://github.com/Expensify/App/blob/5e464eaa3e7614e6f17007e05203e184c275faf2/src/pages/ConciergePage.tsx#L35) with 
```javascript
useEffect(() => {
    if (!isFocused || isLoadingApp){
        return;
    }

    if (session && 'authToken' in session) {
        App.confirmReadyToOpenApp();
        // Pop the concierge loading page before opening the concierge report.
        Navigation.isNavigationReady().then(() => {
            if (isUnmounted.current) {
                return;
            }
            Report.navigateToConciergeChat(true, () => !isUnmounted.current);
        });
    } else {
        Navigation.navigate();
    }

}, [isFocused, isLoadingApp, isUnmounted, session]);
```

<details>
<summary>result</summary>


https://github.com/Expensify/App/assets/41129870/dd74e8da-8eab-447b-a442-1d098d064e47



</details>


But we still have another issue if we open new account with any initial URL, app will open the initial URL util user complete Onboarding then force navigate to concierge/Expensify [here](https://github.com/Expensify/App/blob/fd4987fca779837cb0823091d547625c510b03d8/src/pages/OnboardingPersonalDetails/BaseOnboardingPersonalDetails.tsx#L74-L78) and ignore the initial URL.

Plus, we should fix this also https://github.com/Expensify/App/issues/44235#issuecomment-2201087662

**Comment by NJ-2020** [comment]
Many Thanks, will look into it.

**Comment by NJ-2020** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/44235#issuecomment-2194648301)

**Comment by NJ-2020** [comment]
@ahmedGaber93

Kindly, please take a look at my updated proposal

> But we still have another issue if we open new account with any initial URL, app will open the initial URL util user complete Onboarding then force navigate to concierge/Expensify [here](https://github.com/Expensify/App/blob/fd4987fca779837cb0823091d547625c510b03d8/src/pages/OnboardingPersonalDetails/BaseOnboardingPersonalDetails.tsx#L74-L78) and ignore the initial URL.

This will be fixed in #44260

> In this issue case, The optimistic report already removed, but the app didn't redirect to pre-exist chat because this [condition](https://github.com/Expensify/App/blob/ad2b1926b5d68fd42f7a3e6900258d67ec814858/src/libs/actions/Report.ts#L1300) will return false as the current route now is onboarding/purpose, and the removed report will cause the error page.

We can handle it by checking if the user has not completed the onboarding, and then we can check the previous route before we get redirected into onboarding if the user still looking at the optimistically created report

But as you know if we navigate to a pre-existing chat (which is concierge) it will dismiss and close the onboarding flow

**Comment by NJ-2020** [comment]
> (ConciergePage screen is focused and app is loaded) will be false until onboarding/purpose close and ConciergePage come back to focus.

And also we don't need to use the `useIsFocused` hook, because the useFocusEffect only runs if the screen is currently focused

**Comment by ahmedGaber93** [comment]
I will review this tomorrow in the morning

**Comment by ahmedGaber93** [comment]
@NJ-2020 Could you please add your code changes in your proposal not in the recording video.

**Comment by NJ-2020** [comment]
Done

**Comment by ahmedGaber93** [comment]
> the concierge page sets the isUnmounted to mounted, then unmounted, then mounted again

@NJ-2020 Do you know why concierge page unmount the mount again?

This makes `useEffect` clean up to run while the component is still active


https://github.com/Expensify/App/assets/41129870/7a8764a6-2d22-43aa-96e3-ba110d301a0c

**Comment by NJ-2020** [comment]
Thank you @ahmedGaber93  and I will look into again.

**Comment by NJ-2020** [comment]
> This makes useEffect clean up to run while the component is still active

@ahmedGaber93 This is because of Strict Mode and only applies to development as mentioned in the docs:

> React 18 introduces a new development-only check to Strict Mode. This new check will automatically unmount and remount every component, whenever a component mounts for the first time, restoring the previous state on the second mount.

**Comment by NJ-2020** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/44235#issuecomment-2194648301)

**Comment by ahmedGaber93** [comment]
@NJ-2020 Yeah, that's correct, but I don't think we need to set `isUnmounted.current = true;` in useCallback clean up because it is not designed to call when component unmount see last two lines [here](https://reactnavigation.org/docs/use-focus-effect/), we should instead edit [useEffect](https://github.com/Expensify/App/blob/5ab6be6ecb01d6398eb2503dbd3f6dad60d6638f/src/pages/ConciergePage.tsx#L50-L55) to 
```javascript
useEffect(() => {
    isUnmounted.current = false;

    return () => {
        isUnmounted.current = true;
    }
}, []);
```
and useCallback should be
```javascript
useFocusEffect(
    useCallback(() => {
        if (session && 'authToken' in session) {
            App.confirmReadyToOpenApp();
            // Pop the concierge loading page before opening the concierge report.
            Navigation.isNavigationReady().then(() => {
                if (isUnmounted.current || isLoadingApp) {
                    return;
                }
                Report.navigateToConciergeChat(true, () => !isUnmounted.current);
            });
        } else {
            Navigation.navigate();
        }
    }, [session, isUnmounted, isLoadingApp])
);
```

**Comment by NJ-2020** [comment]
I think the above useEffect is not needed because the useEffect doesn't get unmounted when we change the screen because it only runs one time when we open the Concierge page

### Solution:
We can still use useEffect to get unmounted when navigating to another screen by using `navigation.addListener('blur')` when unmounted and `navigation.addListener('focus')` when mounted

### Here's the code:
```ts
const styles = useThemeStyles();
const [isLoadingApp] = useOnyx(ONYXKEYS.IS_LOADING_APP);
const navigation = useNavigation();
const isUnmounted = useRef(false);

useFocusEffect(
    useCallback(() => {
        if (session && 'authToken' in session) {
            App.confirmReadyToOpenApp();
            // Pop the concierge loading page before opening the concierge report.
            Navigation.isNavigationReady().then(() => {
                if (isUnmounted.current || isLoadingApp) {
                    return;
                }
                Report.navigateToConciergeChat(true, () => !isUnmounted.current);
            });
        } else {
            Navigation.navigate();
        }
    }, [isUnmounted, session, isLoadingApp]),
);

useEffect(() => {
    navigation.addListener('blur', () => {
        isUnmounted.current = true;
    });
    navigation.addListener('focus', () => {
        isUnmounted.current = false;
    });
}, [navigation]);
```

**Comment by NJ-2020** [comment]
@ahmedGaber93 I think the above useEffect is not needed because the useEffect doesn't get unmounted when we change the screen because it only runs one time when we open the Concierge page

### Solution:
We can still use useEffect to get unmounted when navigating to another screen by using `navigation.addListener('blur')` when unmounted and `navigation.addListener('focus')` when mounted

### Here's the code:
```ts
const styles = useThemeStyles();
const [isLoadingApp] = useOnyx(ONYXKEYS.IS_LOADING_APP);
const navigation = useNavigation();
const isUnmounted = useRef(false);

useFocusEffect(
    useCallback(() => {
        if (session && 'authToken' in session) {
            App.confirmReadyToOpenApp();
            // Pop the concierge loading page before opening the concierge report.
            Navigation.isNavigationReady().then(() => {
                if (isUnmounted.current || isLoadingApp) {
                    return;
                }
                Report.navigateToConciergeChat(true, () => !isUnmounted.current);
            });
        } else {
            Navigation.navigate();
        }
    }, [isUnmounted, session, isLoadingApp]),
);

useEffect(() => {
    navigation.addListener('blur', () => {
        isUnmounted.current = true;
    });
    navigation.addListener('focus', () => {
        isUnmounted.current = false;
    });
}, [navigation]);
```

**Comment by ahmedGaber93** [comment]
`isUnmounted.current` should be true when component destroyed, not when blurred.
@NJ-2020 Can you please update your proposal with the code [here](https://github.com/Expensify/App/issues/44235#issuecomment-2211954615)

**Comment by NJ-2020** [comment]
I have looked into it and as far as I understand, we cannot set the `isUnmounted.current` to true when the component is destroyed because, even though we switch screens, the other screen is not unmounted - it's still in memory but not visible. This is a design decision of react-navigation intended for better performance. 

We can only unmount the component by using:
1. `.replace` instead of `.navigate` when navigating to onboarding so it will unmount the component, but I think it needs a lot of changes and is not a good solution
2. We can use this code: 
```ts 
<Tab.Screen
    name="ComponentName"
    component={ComponentName}
    options={{unmountOnBlur: true}}
    listeners={({navigation}) => ({blur: () => navigation.setParams({screen: undefined})})}
/>
```
But this only works in TabNavigator, and not in StackNavigator as mentioned [here](https://github.com/react-navigation/react-navigation/issues/10542#issuecomment-1987361869)

Please kindly let me know your thoughts

**Comment by NJ-2020** [comment]
> @NJ-2020 Can you please update your proposal with the code https://github.com/Expensify/App/issues/44235#issuecomment-2211954615

I've not yet updated the code, I'm waiting for your comment first

**Comment by NJ-2020** [comment]
@ahmedGaber93 Sorry I have changed my [comment](https://github.com/Expensify/App/issues/44235#issuecomment-2212376551) again, and above is my latest update

**Comment by NJ-2020** [comment]
> @NJ-2020 Can you please update your proposal with the code https://github.com/Expensify/App/issues/44235#issuecomment-2211954615

Done, I've updated the proposal

**Comment by ahmedGaber93** [comment]
I am close to accepting the above proposal. We just still have small point in discussion

**Comment by ahmedGaber93** [comment]
I am not able to reproduce the issue in the latest main, but I faced another issue for navigation after complete onboarding, the concierge still loading and not navigate, but I think it may fix here https://github.com/Expensify/App/pull/44861. So we can hold this for https://github.com/Expensify/App/pull/44861 then recheck




https://github.com/Expensify/App/assets/41129870/ebebd79a-e2cc-4acf-b61d-080273586abc

**Comment by NJ-2020** [comment]
Okay

**Comment by isabelastisser** [comment]
@ahmedGaber93, to clarify, is this still on hold? 

> I think it may fix here https://github.com/Expensify/App/pull/44861. So we can hold this for https://github.com/Expensify/App/pull/44861 then recheck

**Comment by madmax330** [comment]
Still on hold

**Comment by madmax330** [comment]
https://github.com/Expensify/App/pull/44861 is still not merged

**Comment by ahmedGaber93** [comment]
https://github.com/Expensify/App/pull/44861 is still not merged

**Comment by ahmedGaber93** [comment]
#44861 is still not merged

**Comment by ahmedGaber93** [comment]
https://github.com/Expensify/App/pull/44861 is still not merged

**Comment by NJ-2020** [comment]
#44861 got merged
@ahmedGaber93 When trying to reproduce the issue again, the concierge page shows infinite loading and does not navigate.
This [issue](https://github.com/Expensify/App/issues/44235#issuecomment-2221411523) is still reproducible

**Comment by ahmedGaber93** [comment]
checking

**Comment by ahmedGaber93** [comment]
> When trying to reproduce the issue again, the concierge page shows infinite loading and does not navigate.

@NJ-2020 Yeah, but I think the issue now is strict mode related as you explained here https://github.com/Expensify/App/issues/44235#issuecomment-2211621919, and can fix only by the below `useEffect`, Can you confirm this?
```javascript
useEffect(() => {
    isUnmounted.current = false;

    return () => {
        isUnmounted.current = true;
    }
}, []);
```

**Comment by NJ-2020** [comment]
@ahmedGaber93 Yes we can

**Comment by NJ-2020** [comment]
@ahmedGaber93 Done, Proposal updated

**Comment by NJ-2020** [comment]
@ahmedGaber93 [PR](https://github.com/Expensify/App/pull/46483) is ready for review

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by ahmedGaber93** [comment]
We have a different issue, and on dev only. PR now in review.

**Comment by isabelastisser** [comment]
@ahmedGaber93 

> We have a different issue, and on dev only. PR now in review.

Does this mean that this issue can be closed?

**Comment by ahmedGaber93** [comment]
@isabelastisser No, keep it open. we are still working on it to match the expected behavior

**Comment by isabelastisser** [comment]
Payment summary:

$250 to @ahmedGaber93 for C+ review. Paid in Upwork.
@250 to @NJ-2020 for proposal. Paid in Upwork.

**Comment by isabelastisser** [comment]
@ahmedGaber93, I DM'd you for visibility to complete the BZ list. Thanks!

**Comment by ahmedGaber93** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [x] [@ahmedGaber93] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/42592
- [x] [@ahmedGaber93] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/42592#issuecomment-2299229031
- [x] [@ahmedGaber93] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
- [x] [@ahmedGaber93] Determine if we should create a regression test for this bug. N
- [x] [@ahmedGaber93] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

#### this a DEV only issue, so I think no need to create a regression test.

**Comment by isabelastisser** [comment]
Thanks, @ahmedGaber93! All set!
